### PR TITLE
feat(Core/Scripts): Optimize PlayerScripts by calling only overridden/implemented functions.

### PR DESCRIPTION
**Pull request by walkline**
## Changes Proposed:

At the moment, our scripting system will try to call all available hooks for a registered script. So, if you created and registered a player script that implements only the OnLogin hook, the core would try to call all other possible hooks for this script, such as OnBeforePlayerUpdate, OnPlayerUpdate, OnIsPvP, etc. (but it would execute dummy hooks). And the more scripts you register, the greater the performance impact.

With the suggested changes for PlayerScripts, we would call only hooks that are enabled.
To use this optimization, PlayerScripts should add additional parameter during script initialization. For example
```
class ServerMailReward : public PlayerScript
{
public:
    ServerMailReward() : PlayerScript("ServerMailReward", {PLAYERHOOK_ON_LOGIN}) { }

    void OnLogin(Player* player) override
    {
```

If you don't provide this parameter, then all hooks would be enabled for the script, which makes the change backward compatible.

I'm a little bit exhausted to write boilerplate code, so feel free to update other scriptable objects (like UnitScript, GuildScript, etc.) by utilizing this approach.

This PR proposes changes to:
-  [x] Core (units, players, creatures, game systems).
-  [x] Scripts (bosses, spell scripts, creature scripts).
-  [ ] Database (SAI, creatures, etc).

## Issues Addressed:

- Closes 

## SOURCE:

The changes have been validated through:
- [ ] Live research (checked on live servers, e.g Classic WotLK, Retail, etc.)
- [ ] Sniffs (remember to share them with the open source community!)
- [ ] Video evidence, knowledge databases or other public sources (e.g forums, Wowhead, etc.)
- [ ] The changes promoted by this pull request come partially or entirely from another project (cherry-pick). **Cherry-picks must be committed using the proper --author tag in order to be accepted, thus crediting the original authors, unless otherwise unable to be found**

## Tests Performed:

This PR has been:
- [ ] Tested in-game by the author.
- [ ] Tested in-game by other community members/someone else other than the author/has been live on production servers.
- [ ] This pull request requires further testing and may have edge cases to be tested.

## How to Test the Changes:

- [ ] This pull request can be tested by following the reproduction steps provided in the linked issue
- [ ] This pull request requires further testing. Provide steps to test your changes. If it requires any specific setup e.g multiple players please specify it as well.

1.
2.
3.

## Known Issues and TODO List:

- [ ]
- [ ]

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by sudlud** [comment]
Nice

**Comment by pangolp** [comment]
After this change, we may have to update the modules, right?

**Comment by heyitsbench** [comment]
We may want to update the modules, but it wouldn't be required for them to continue to function.

**Comment by pangolp** [comment]
Update the branch, with the master branch, because some changes to the tests were added. You should pass the tests now.

**Comment by sudlud** [comment]
What's the performance improvement on clean AC with this patch? :) did you do any tests already, just wondering

**Comment by walkline** [comment]
@sudlud In my case, with around 1k idle players, I see approximately a 11% performance improvement. With real players, I believe the numbers would be different.

**Comment by sudlud** [comment]
@walkline alright, thanks for the feedback

**Comment by kissingers** [comment]
After update, will it still work if the original module not change?

**Comment by walkline** [comment]
@kissingers Yes, they should work.

**Comment by PkllonG** [comment]
I thought about it for a while, and I was afraid that you would be unhappy when you heard my opinion.
I'm sorry I haven't been testing this PR, Because
If I had to choose, I'd rather remove the unused hooks.

**Comment by sudlud** [comment]
@PkllonG removing is not really an option imo

Well, the PR actually "removes" the unused hooks of each PlayerScript instance from the to-execute-list, so the unused hooks of each script don't get called in runtime.

E.g. core already has at least 4 playerscripts which only each use 1-2 hooks actively and the PR disables the rest of the 200 for each script.

But just because the core is not using all 200 hooks of PlayerScript doesn't mean they have no purpose at all. Modules also use scripts and hooks.

**Comment by PkllonG** [comment]
> @PkllonG removing is not really an option imo
> 
> Well, the PR actually "removes" the unused hooks of each PlayerScript instance from the to-execute-list, so the unused hooks of each script don't get called in runtime.
> 
> E.g. core already has at least 4 playerscripts which only each use 1-2 hooks actively and the PR disables the rest of the 200 for each script.
> 
> But just because the core is not using all 200 hooks of PlayerScript doesn't mean they have no purpose at all. Modules also use scripts and hooks.
> 
> Edit: The whole purpose of hooks is that they exist if anyone wants to use them, but they should have zero impact on performance if unused.

Well, it may be that everyone's choice is not in the same direction. I understand how PR works, and thank you for your careful explanation.

**Comment by sudlud** [comment]
@PkllonG  Thanks for your feedback!

Would you still be willing to test the PR? 🙂 

Your frequent testing has been quite valuable on recent PRs.

**Comment by PkllonG** [comment]
> @PkllonG Thanks for your feedback!
> 
> Would you still be willing to test the PR? 🙂
> 
> Your frequent testing has been quite valuable on recent PRs.

okey.

**Comment by PkllonG** [comment]
With or without it ```std::vector<uint16>{PLAYERHOOK_ON_LOGIN}``` , module works fine.

**Comment by PkllonG** [comment]
As for performance. I didn't test on the server.

**Comment by sudlud** [comment]
@PkllonG thanks for testing!

@walkline do you want to merge master branch into PR to get all checks passing?

**Comment by pangolp** [comment]
I don't know if it will pass the tests, because in the core, a series of modules are also validated. I just made a pull request, so that the emulator stops validating those modules. The modules already have tests. It is not necessary for the emulator to validate them every time someone wants to make a contribution. There are people who do not use modules, who use the clean core, blizzlike. https://github.com/azerothcore/azerothcore-wotlk/pull/18690

**Comment by sudlud** [comment]
I tested this PR in regards of the PlayerScript in midsummer.cpp.

The `OnUpdateZone`-Hook still works as expected.

I've then updated the hook list, replacing the hook with a random not-implemented hook of the PlayerScript:
```diff
-MidsummerPlayerScript() : PlayerScript("MidsummerPlayerScript", std::vector<uint16>{PLAYERHOOK_ON_UPDATE_ZONE})
+MidsummerPlayerScript() : PlayerScript("MidsummerPlayerScript", std::vector<uint16>{PLAYERHOOK_ON_LOGIN})
```

The `OnUpdateZone`-Hook was not beeing called anymore (as expected).

I then removed the hook list (to check backwards-compatibility):
```diff
-MidsummerPlayerScript() : PlayerScript("MidsummerPlayerScript", std::vector<uint16>{PLAYERHOOK_ON_UPDATE_ZONE})
+MidsummerPlayerScript() : PlayerScript("MidsummerPlayerScript")
```

The `OnUpdateZone`-Hook works as expected (again).

So I couldn't find any issues, PR works as expected (for this test scenario / PlayerScript).

**Review by Nyeriah — APPROVED** [review]

File: src/server/game/Scripting/ScriptMgr.h [hunk 1]
```diff
@@ -710,30 +710,42 @@ class ScriptRegistry
...
     // after server startup.
     static ScriptMap ScriptPointerList;
     // After database load scripts
     static ScriptVector ALScripts;
+    // The list of hook types with the list of enabled scripts for this specific hook.
+    // With this approach, we wouldn't call all available hooks in case if we override just one hook.
+    static EnabledHooksVector EnabledHooks;
 
-    static void AddScript(TScript* const script)
+    static void AddScript(TScript* const script, std::vector<uint16> enabledHooks = std::vector<uint16>())
```
Comment (Winfidonarleyan) [inline src/server/game/Scripting/ScriptMgr.h:728]:
```suggestion
    static void AddScript(TScript* const script, std::vector<uint16> enabledHooks = {})
```

File: src/server/game/Scripting/ScriptMgr.h [hunk 2]
```diff
@@ -710,30 +710,42 @@ class ScriptRegistry
...
     {
         ASSERT(script);
 
         if (!_checkMemory(script))
             return;
 
+        if (EnabledHooks.size() == 0)
```
Comment (Winfidonarleyan) [inline src/server/game/Scripting/ScriptMgr.h:735]:
```suggestion
        if (EnabledHooks.empty())
```

File: src/server/game/Scripting/ScriptMgr.h [hunk 3]
```diff
@@ -710,30 +710,42 @@ class ScriptRegistry
...
+            EnabledHooks.resize(script->GetTotalAvailableHooks());
```
Comment (Winfidonarleyan) [inline src/server/game/Scripting/ScriptMgr.h:736]:
```suggestion
            EnabledHooks.reserve(script->GetTotalAvailableHooks());
```

File: src/server/game/Scripting/ScriptMgr.h [hunk 4]
```diff
@@ -710,30 +710,42 @@ class ScriptRegistry
...
+
         if (script->isAfterLoadScript())
         {
-            ALScripts.push_back(script);
+            ALScripts.push_back(std::pair<TScript*,std::vector<uint16>>(script, enabledHooks));
```
Comment (Winfidonarleyan) [inline src/server/game/Scripting/ScriptMgr.h:740]:
```suggestion
            ALScripts.emplace_back(script, std::move(enabledHooks));
```

File: src/server/game/Scripting/ScriptMgr.h [hunk 5]
```diff
@@ -776,6 +788,14 @@ class ScriptRegistry
                     // If the script is already assigned -> delete it!
                     if (oldScript)
                     {
+                        for (std::vector<TScript*>& vIt : EnabledHooks)
```
Comment (Winfidonarleyan) [inline src/server/game/Scripting/ScriptMgr.h:791]:
```suggestion
                        for (auto& vIt : EnabledHooks)
```

File: src/server/game/Scripting/ScriptMgr.h [hunk 6]
```diff
@@ -798,6 +818,9 @@ class ScriptRegistry
             }
             else
             {
+                for (uint16 v : (*it).second)
+                    EnabledHooks[v].push_back(script);
```
Comment (Winfidonarleyan) [inline src/server/game/Scripting/ScriptMgr.h:822]:
```suggestion
                    EnabledHooks[v].emplace_back(script);
```

File: src/server/game/Scripting/ScriptObject.h [hunk 7]
```diff
@@ -53,15 +53,18 @@ class ScriptObject
 
     [[nodiscard]] const std::string& GetName() const { return _name; }
 
+    uint16 GetTotalAvailableHooks() { return _totalAvailableHooks; }
```
Comment (Winfidonarleyan) [inline src/server/game/Scripting/ScriptObject.h:56]:
```suggestion
    [[nodiscard]] uint16 GetTotalAvailableHooks() { return _totalAvailableHooks; }
```

File: src/server/game/Scripting/ScriptDefines/PlayerScript.cpp [hunk 8]
```diff
@@ -21,1736 +21,1228 @@
 
 void ScriptMgr::OnBeforePlayerDurabilityRepair(Player* player, ObjectGuid npcGUID, ObjectGuid itemGUID, float& discountMod, uint8 guildBank)
 {
-    ExecuteScript<PlayerScript>([&](PlayerScript* script)
-    {
+    CALL_ENABLED_HOOKS(PlayerScript, PLAYERHOOK_ON_BEFORE_DURABILITY_REPAIR, {
         script->OnBeforeDurabilityRepair(player, npcGUID, itemGUID, discountMod, guildBank);
```
Comment (Winfidonarleyan) [inline src/server/game/Scripting/ScriptDefines/PlayerScript.cpp:25]:
```suggestion
    CALL_ENABLED_HOOKS(PlayerScript, PLAYERHOOK_ON_BEFORE_DURABILITY_REPAIR,
    {
        script->OnBeforeDurabilityRepair(player, npcGUID, itemGUID, discountMod, guildBank);
```

File: src/server/game/Scripting/ScriptMgr.h (hunk 3 above)
Comment (walkline) [reply src/server/game/Scripting/ScriptMgr.h:736]: Here we need resize, since later we are accessing elements like this: `EnabledHooks[v].`

Comment (Winfidonarleyan) [reply src/server/game/Scripting/ScriptMgr.h:736]: Ah, yes, my bad

**Review by Winfidonarleyan — APPROVED** [review]

File: src/server/scripts/World/server_mail.cpp [hunk 9]
```diff
@@ -24,7 +24,7 @@
 class ServerMailReward : public PlayerScript
 {
 public:
-    ServerMailReward() : PlayerScript("ServerMailReward") { }
+    ServerMailReward() : PlayerScript("ServerMailReward", {PLAYERHOOK_ON_FIRST_LOGIN}) { }
```
Comment (PkllonG) [inline src/server/scripts/World/server_mail.cpp:27]: PLAYERHOOK_ON_LOGIN

**Comment by pangolp** [comment]
Excellent work, thank you very much.

**Comment by PkllonG** [comment]
nice!

**Review by PkllonG — APPROVED** [review]
Great job

**Comment by pangolp** [comment]
@walkline maybe I can annoy you a little bit, let's see if we can write something on the wiki about this (if necessary). I asked if we had to update the modules, but not as a criticism, but as a question, if we really had to do it or not, and it would be interesting, if we have to do it, we can give some example just in case or something.

**Pull request merged**